### PR TITLE
Update halo2 dependency to v2023_04_20

### LINKED ISSUE #1366: Upgrade Halo2 version to v2023_04_20
### Describe the feature you would like

https://github.com/privacy-scaling-explorations/halo2/releases/tag/v2023_04_20

### Additional context

Unblock #1220

### PR DESCRIPTION
**Pull request by ed255**
### Description

Update halo2 to the latest dependency, which contains some changes affecting the `ff` trait.

### Issue Link

Resolve #1366

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Rationale

I've extended the `Field` trait we define in `eth_types` to contain the common traits used in different places, to avoid having a custom list of traits in each function.

### CODE REVIEW

**Comment by CPerezz** [comment]
Need help with that?

**Comment by ed255** [comment]
> Need help with that?

I think I do :smiling_face_with_tear:  
So far I have solved the failures in the tx_circuit.  They were unrelated to cell re-assignment.
I've also checked the evm_circuit failure.  So here's a summary of where we are:

In https://github.com/privacy-scaling-explorations/halo2/pull/129 Han introduced a check for the mock prover advice assignment with the following comment: `Inconsistent assignment between different phases.`.  As far as I see the check also applies when doing inconsistent assignment in the same phase.  Now we have:

- evm_circuit: The assignment is coming from StoredExpression / CachedRegion. To assign each execution step, we do an optimistic assignment of the next step because sometimes expressions in the current step need values of the next one.  I say optimistic because there are some values of the next step that are not known yet, so we first do an assignment that may not be correct, and then re-do it with the correct values.
- state_circuit: Negative tests do re-assignment of different values on purpose, so it's not clear how this can be fixed

Both cases seem reasonable to me, and they are not friendly to the re-assignment with different value check.  So I would like help on ideas to proceed!

I have thought about using the Cached region to first store all the assignments (where we can re-assign freely), and after everything is assigned, we do the real assignment.  I think this would add a lot of unneeded complexity.  Moreover, the assignment function returns an AssignedCell which can't be constructed outside of halo2

The other option that comes to my mind is to disable the re-assignment check, maybe with a toggle in the MockProver, or something like that, and then we can disable it for the EVM and state circuits tests.  Nevertheless that means we lose the original feature.

A final option is to redo the feature to make the MockProver complain on reassignment with different values, **only** when the new assignment is in a different phase.  For that we would need to store the phase of each assignment, which would add some overhead.

**Comment by CPerezz** [comment]
> I have thought about using the Cached region to first store all the assignments (where we can re-assign freely), and after everything is assigned, we do the real assignment. I think this would add a lot of unneeded complexity. Moreover, the assignment function returns an AssignedCell which can't be constructed outside of halo2

Agree this is not a way.

> The other option that comes to my mind is to disable the re-assignment check, maybe with a toggle in the MockProver, or something like that, and then we can disable it for the EVM and state circuits tests. Nevertheless that means we lose the original feature.

A final option is to redo the feature to make the MockProver complain on reassignment with different values, only when the new assignment is in a different phase. For that we would need to store the phase of each assignment, which would add some overhead.

Do we really want the check?? It just seems there's no good solution for the state circuit case. Let's just get ridd of it for now?

File: Cargo.toml [hunk 1]
```diff
@@ -13,8 +13,13 @@ members = [
...
 
-[patch.crates-io]
-halo2_proofs = { git = "https://github.com/privacy-scaling-explorations/halo2.git", tag = "v2023_02_02" } 
+# Uncomment after the next halo2 release
+# [patch.crates-io]
+# halo2_proofs = { git = "https://github.com/privacy-scaling-explorations/halo2.git", tag = "v2023_04_20" }
+
+# Remove after the next halo2 release
+[patch."https://github.com/privacy-scaling-explorations/halo2.git"]
+halo2_proofs = { git = "https://github.com/appliedzkp/halo2.git", rev = "be955686f86eb618f55d2320c0e042485b313d22" }
```
Comment (han0110) [inline Cargo.toml:22]: Should we just override `v2023_04_20` with current latest halo2, or should we create a new release (I'd prefer the former for easier management for us)?

**Review by han0110 — COMMENTED** [review]
LGTM! Nice work! Just some nits that could be ignored, and also a question about halo2 dependency.

File: gadgets/src/less_than.rs [hunk 2]
```diff
@@ -110,7 +109,7 @@ impl<F: Field, const N_BYTES: usize> LtChip<F, N_BYTES> {
     }
 }
 
-impl<F: Field, const N_BYTES: usize> LtInstruction<F> for LtChip<F, N_BYTES> {
+impl<F: Field + Ord, const N_BYTES: usize> LtInstruction<F> for LtChip<F, N_BYTES> {
```
[the remark this answers is not in the document]
Comment (ed255) [reply gadgets/src/less_than.rs]: Thanks for catching these unneeded `+ Ord`, I've removed them all.

File: zkevm-circuits/src/bytecode_circuit/test.rs [hunk 3]
```diff
@@ -6,7 +6,7 @@ use crate::{
 };
 use bus_mapping::evm::OpcodeId;
 use eth_types::{Bytecode, Field, Word};
-use halo2_proofs::{dev::MockProver, halo2curves::bn256::Fr};
+use halo2_proofs::{arithmetic::Field as Halo2Field, dev::MockProver, halo2curves::bn256::Fr};
```
[the remark this answers is not in the document]
Comment (ed255) [reply zkevm-circuits/src/bytecode_circuit/test.rs:9]: > Oh, I see the collision between `eth_types` and `arithmetic`, Never mind sweat_smile

That's it!  We need to import the trait to use it's methods, but at the same time it collides with `eth_types::Field` :P

**Pull request merged**